rust/xaynet-sdk/src/state_machine/phases/update.rs

```diff
@@ -0,0 +1,224 @@
+use std::ops::Deref;
+
+use derive_more::From;
+use xaynet_core::{
+    crypto::Signature,
+    mask::{MaskObject, MaskSeed, Masker, Model},
+    message::Update as UpdateMessage,
+    LocalSeedDict,
+    ParticipantTaskSignature,
+    SumDict,
+};
+
+use crate::{
+    state_machine::{Phase, Progress, Step, TransitionOutcome, IO},
+    MessageEncoder,
+};
+
+#[derive(From)]
+pub enum LocalModel {
+    Dyn(Box<dyn AsRef<Model> + Send>),
+    Owned(Model),
+}
+
+impl AsRef<Model> for LocalModel {
+    fn as_ref(&self) -> &Model {
+        match self {
+            LocalModel::Dyn(model) => model.deref().as_ref(),
```

Comment:
I thought `deref` would fit well because the box is a pointer that we're dereferencing. But yeah they are stictly equivalent. I'm wondering if I could use only one or if `(*model).as_ref()` would work. But I think the `*` operator tries to move out of the box no?

Reply:
afaik, the `*` is just syntactic sugar for `deref()`, so `(*model).as_ref()` should work as well. this is probably the most concise and descriptive expression of the three.

if you have a `x: &T` and call `*x`, then it should move. but if you call `&*x` then it should just give a reference without moving. similarly for a box.

Reply:
somehow that doesn't work:

```
error[E0308]: mismatched types
  --> xaynet-sdk/src/state_machine/phases/update.rs:27:39
   |
27 |             LocalModel::Dyn(model) => (*model).as_ref(),
   |                                       ^^^^^^^^^^^^^^^^^ expected struct `Model`, found trait object `dyn AsRef`
   |
   = note: expected reference `&Model`
              found reference `&dyn AsRef<Model> + Send`
```

Reply:
that's indeed a tricky one. the issue comes from the function signature, since `as_ref(&self)` will cause the `self` in the `match` to be of type `&LocalModel` instead of `LocalModel`. hence the bound variable `model` is of type `&Box<dyn AsRef<Model> + Send>` instead of `Box<dyn AsRef<Model> + Send>`.
then the `*` operator only strips of the first `&`, but not the `Box`, and the final `as_ref()` will turn the `Box` into a `&` giving a `&dyn AsRef<Model> + Send`.

to fix this with the `*` operator, we would need to use it twice: `(**model).as_ref()`. what could also work is to write the match statement as `match &self`, then `self` here should be of type `LocalModel` instead of `&LocalModel`.

i was wondering why it works with `deref()` in the first place. because `model` is of type `&Box<dyn AsRef<Model> + Send>` the call to `deref()` will return a `&dyn AsRef<Model> + Send`, which can then in turn be consumed by the `as_ref()` properly.